### ISSUE TITLE
Coinduction incompatible with univalence  [WAS: Structural termination order incompatible with univalence]

### ISSUE THREAD

**Issue by GoogleCodeExporter** [agda/agda#1023]
Following discussions on the coq-club it was discovered that Agda's current 
--without-K does not guarantee compatibility with univalence.  My version of 
--without-K DOES rule out the bad use of Refl in the following test case, an 
argument for accepting my patch (see issue 865).
```agda
-- Andreas, 2014-01-10
-- Code by Jesper Cockx and Conor McBride and folks from the Coq-club

{-# OPTIONS --without-K #-}

-- An empty type.

data Zero : Set where

-- A unit type as W-type.

mutual
  data WOne : Set where wrap : FOne -> WOne
  FOne = Zero -> WOne

-- Type equality.

data _<->_ (X : Set) : Set -> Set where
  Refl : X <-> X

-- This postulate is compatible with univalence:

postulate
  iso : WOne <-> FOne

-- But accepting that is incompatible with univalence:

noo : (X : Set) -> (WOne <-> X) -> X -> Zero
noo .WOne Refl (wrap f) = noo FOne iso f

-- Matching against Refl silently applies the conversion
-- FOne -> WOne to f.  But this conversion corresponds
-- to an application of wrap.  Thus, f, which is really
-- (wrap f), should not be considered a subterm of (wrap f)
-- by the termination checker.
-- At least, if we want to be compatible with univalence.

absurd : Zero
absurd = noo FOne iso (\ ())

```

Original issue reported on code.google.com by `andreas....@gmail.com` on 10 Jan 2014 at 8:27

**Comment by Saizan** [comment]
It seems that the check for the original problem can be circumvented by the identity type former as a datatype.

The following proof of false typechecks on Agda version 2.6.2-c857009

```agda

{-# OPTIONS --without-K #-}

-- An empty type.

data Zero : Set where

-- A unit type as W-type.

mutual
  data WOne : Set where wrap : FOne -> WOne
  FOne = Zero -> WOne

-- Type equality.

data _<->_ (X : Set) : Set -> Set₁ where
  Refl : X <-> X

-- This postulate is compatible with univalence:

postulate
  iso : WOne <-> FOne

-- But accepting that is incompatible with univalence:

data Id (X : Set) : Set where
  i : X → Id X

noo : (X : Set) -> (WOne <-> X) -> Id X -> Zero
noo .WOne Refl (i (wrap f)) = noo FOne iso (i f)

absurd : Zero
absurd = noo FOne iso (i \ ())

```

**Comment by jespercockx** [comment]
It seems I do not understand how the termination checker works, as I would not have expected this definition to pass the termination checker. Why is it fine to consider `i f` as being structurally smaller than `i (wrap f)`? (EDIT: I mean, I don't even understand why this code is fine `--with-K`.)

**Comment by jespercockx** [comment]
I re-checked our papers on pattern matching without K, and I'm pretty sure there's no way this example would be accepted by the (rather simplistic) termination criterion we're using there. So this seems to be caused by insufficient integration between the theories supporting the different features of Agda (in this case: pattern matching and termination checking).

**Comment by Saizan** [comment]
I was also a bit surprised by `i f` being smaller, maybe it's size-changing termination?, but we can rework the example to not rely on it:
```agda
{-# OPTIONS --without-K #-}

data Zero : Set where

data Id (X : Set) : Set where
  i : X → Id X

mutual
  data WOne : Set where wrap : Id FOne -> WOne
  FOne = Zero -> WOne

data _<->_ (X : Set) : Set -> Set₁ where
  Refl : X <-> X

postulate
  iso : WOne <-> FOne

mutual
  noo : (X : Set) -> (WOne <-> X) -> Id X -> Zero
  noo .WOne Refl (i (wrap f)) = noo FOne iso f

absurd : Zero
absurd = noo FOne iso (i \ ())
```

**Comment by Saizan** [comment]
Do we know how this problem was addressed in Coq?

**Comment by jespercockx** [comment]
I believe in Coq, constructor arguments are marked as either "recursive" or "non-recursive", and the termination checker only accepts recursive calls on arguments that come from a recursive position. In this example the argument of `i` would not be seen as recursive, so the recursive call would be rejected.

However, adopting such a criterion in Agda might also rule out a lot of "good" use cases, e.g. where one first defines a (non-recursive) functor and then uses a separate fixpoint construction, e.g.
```agda
data Maybe (A : Set) : Set where
  just    : A → Maybe A
  nothing : Maybe A

data Nat : Set where
  inj : Maybe Nat → Nat

id : Nat → Nat
id (inj (just x)) = inj (just (id x))
id (inj nothing)  = inj nothing
```
This is something we probably still want to accept, but (I presume) would be rejected by Coq's approach because the argument to `just` is not recursive.

**Comment by Saizan** [comment]
Is their fix for the original problem of this issue also similar to what we do in Agda? Since that definition of `noo` only matched on `wrap` whose argument is recursive.

**Comment by jespercockx** [comment]
I tried porting the example to Coq, but it seems it's not accepted:
```coq
Inductive Zero : Prop := .

Inductive Id (X : Type) : Type :=
  | i : X -> Id X.

Inductive One : Type :=
  | wrap : Id (Zero -> One) -> One.

Inductive Equiv (X : Type) : Type -> Type :=
  | Refl : Equiv X X.

Axiom iso : Equiv One (Zero -> One).

Fixpoint noo (X : Type) (eq : Equiv One X) (x : Id X) : Zero :=
  (match eq in Equiv _ X return Id X -> Zero with
  | Refl _ => fun x => 
    match (x : Id One) with
    | i _ o => 
      match o with
      | wrap f => noo (Zero -> One) iso f
      end
    end
  end) x.
```
Error message: `Cannot guess decreasing argument of fix.`

I'm not sure what Coq exactly does here to reject the example, it seems that it somehow cares about the first argument of the recursive call to `noo`.

**Comment by jespercockx** [comment]
Thinking a bit more about it, the reason why Coq rejects the example is because I had to abstract over `x : Id X` in order for the pattern match on `Refl` to have the desired effect. If you don't do the abstraction, then the type of `x` remains at `Id X` instead of being specialized to `Id One`. However, that means the link between the argument `f` in the recursive call and the original argument `x` is lost, which means the termination checker rejects the recursive call.

In other words, it seems that the Coq criterion can be summarized as "the type of the argument on which the function is structurally recursive cannot be refined by any pattern matches." Or perhaps a bit more refined: "the type of the argument on which the function is structurally recursive must be informative enough to access the argument used in the recursive call without any refinements from (other) pattern matches."

**Comment by jespercockx** [comment]
Another way to approach the problem is by saying that we can compute in advance an infinite tree of all valid recursive calls from the type of the argument on which we are recursing. The way we construct this infinite tree is just by repeatedly case splitting on a fresh variable of this type and collecting all the possible recursive arguments generated by this process. Crucially, this should be done using the type before any refinements. Some examples:

- For the buildin `Nat` we get the patterns `x`, `suc x`, `suc (suc x)`, ..., saying that we can recurse on arguments of arbitrary depth
- For the definition of `Nat` as the fixpoint of `Maybe` we get `x`, `inj (just x)`, inj (just (inj (just x))`, ..., again allowing us to recurse on arguments of arbitrary depth
- For the original definition of `noo` the argument has type `X`, on which we can do no case splitting so there are no valid recursive calls
- For Andrea's new definition of `noo` the argument has type `Id X`, on which we can match as `i y` but then we're stuck. So again there are no valid recursive calls.

However, using this process naively to check the validity of recursive calls could be prohibitively expensive. Maybe we can devise a cheaper way to approximate it?

**Comment by Saizan** [comment]
Oh, I see, I was under the impression Coq also had this problem, but maybe it actually was only discussed on coq-club. Unfortunately I can't seem to find the coq-club thread.

**Comment by andreasabel** [comment]
Meta-comment: Please do not reopen issues whose fix has been shipped.  Rather, open a new issue linking to the old one:
- #5910
